drivers/timer/realtek_rts5912_rtmr.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2024 Realtek Semiconductor Corporation, SIBG-SD7
+ * Author: Lin Yu-Cheng <lin_yu_cheng@realtek.com>
+ */
+
+#define DT_DRV_COMPAT realtek_rts5912_rtmr
+
+#include <zephyr/init.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/irq.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/drivers/timer/system_timer.h>
+
+#include <reg/reg_rtmr.h>
+#include <reg/reg_system.h>
+
+#define RTS5912_SCCON_REG_BASE ((SYSTEM_Type *)(DT_REG_ADDR(DT_NODELABEL(sccon))))
+
+#define CYCLES_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "Realtek RTOS timer is not supported multiple instances");
+
+#define RTMR_REG ((RTOSTMR_Type *)DT_INST_REG_ADDR(0))
+
+#define SLWTMR_REG                                                                                 \
+	((RTOSTMR_Type *)(DT_REG_ADDR(DT_COMPAT_GET_ANY_STATUS_OKAY(realtek_rts5912_slwtimer))))
+
+#define SSCON_REG ((SYSTEM_Type *)(DT_REG_ADDR(DT_NODELABEL(sccon))))
+
+#define RTMR_COUNTER_MAX   0x0ffffffful
+#define RTMR_COUNTER_MSK   0x0ffffffful
+#define RTMR_TIMER_STOPPED 0xf0000000ul
+
+#define MAX_TICKS ((k_ticks_t)(RTMR_COUNTER_MAX / CYCLES_PER_TICK) - 1)
+
+#define RTMR_ADJUST_LIMIT  2
+#define RTMR_ADJUST_CYCLES 1
+
+static struct k_spinlock lock;
+static uint32_t accumulated_cycles;
+static uint32_t previous_cnt;      /* Record the counter set into RTMR */
+static uint32_t last_announcement; /* Record the last tick announced to system */
+
+static void rtmr_restart(uint32_t counter)
+{
+	RTMR_REG->CTRL = 0ul;
+	RTMR_REG->LDCNT = counter;
+	RTMR_REG->CTRL = RTOSTMR_CTRL_INTEN_Msk | RTOSTMR_CTRL_EN_Msk;
+}
+
+static uint32_t rtmr_get_counter(void)
+{
+	uint32_t counter = RTMR_REG->CNT;
+
+	if ((counter == 0) && (RTMR_REG->CTRL & RTOSTMR_CTRL_EN_Msk)) {
+		counter = previous_cnt;
+	}
+
+	return counter;
+}
+
+static void rtmr_isr(const void *arg)
+{
+	ARG_UNUSED(arg);
+
+	uint32_t cycles;
+	int32_t ticks;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	rtmr_restart(RTMR_COUNTER_MAX * CYCLES_PER_TICK);
+
+	cycles = previous_cnt;
+	previous_cnt = (RTMR_COUNTER_MAX * CYCLES_PER_TICK);
+
+	accumulated_cycles += cycles;
+	accumulated_cycles &= RTMR_COUNTER_MSK;
```

Comment:
bit confused by these masks, when would they apply? Would it make sense to check for `accumulated_cycles > RTMR_COUNTER_MSK` and maybe clamp instead? (sory not super familiar with what this is doing and its corner cases)

Reply:
We use the  `if function` to check whether `accumulated_cycles` is bigger than `RTMR_COUNTER_MSK`.
If it is, we will clamp it into RTMR_COUNTER_MSK range.